src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+	}
+
+	public AssemblyDefinition Resolve (string fullName, ReaderParameters? parameters = null)
+	{
+		return Resolve (AssemblyNameReference.Parse (fullName), parameters);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name)
+	{
+		return Resolve (name, null);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		return Resolve (AndroidTargetArch.None, name, parameters);
+	}
+
+	public AssemblyDefinition Resolve (AndroidTargetArch arch, AssemblyNameReference name, ReaderParameters? parameters = null)
+	{
+		string shortName = name.Name;
+		if (cache.TryGetValue (shortName, out CacheEntry? entry)) {
+			return SelectAssembly (arch, name.FullName, entry, loading: false);
+		}
+
+		if (arch == AndroidTargetArch.None) {
+			return FindAndLoadFromDirectories (arch, FrameworkSearchDirectories, name, parameters);
+		}
+
+		if (!AbiSearchDirectories.TryGetValue (arch, out ICollection<string>? directories) || directories == null) {
+			throw CreateLoadException (name);
+		}
+
+		return FindAndLoadFromDirectories (arch, directories, name, parameters);
+	}
+
+	AssemblyDefinition FindAndLoadFromDirectories (AndroidTargetArch arch, ICollection<string> directories, AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		string? assemblyFile;
+		AssemblyDefinition? candidate = null;
+		foreach (string dir in directories) {
+			if ((assemblyFile = SearchDirectory (name.Name, dir)) != null) {
+				AssemblyDefinition? loaded = Load (arch, assemblyFile);
+				if (Array.Equals (loaded?.Name.MetadataToken, name.MetadataToken)) {
+					return loaded;
+				}
+				candidate = candidate ?? loaded;
+			}
+		}
+
+		if (candidate != null)
+			return candidate;
+
+		throw CreateLoadException (name);
+	}
+
+	static FileNotFoundException CreateLoadException (AssemblyNameReference name)
+	{
+		return new FileNotFoundException ($"Could not load assembly '{name}'.");
+	}
+
+	static string? SearchDirectory (string name, string directory)
+	{
+		if (Path.IsPathRooted (name) && File.Exists (name)) {
+			return name;
+		}
+
+		var file = Path.Combine (directory, $"{name}.dll");
+		if (File.Exists (file)) {
+			return file;
+		}
+
+		return null;
+	}
+
+	public virtual AssemblyDefinition? Load (AndroidTargetArch arch, string filePath, bool forceLoad = false)
+	{
+		string name = Path.GetFileNameWithoutExtension (filePath);
+		AssemblyDefinition? assembly;
+		if (!forceLoad && cache.TryGetValue (name, out CacheEntry? entry)) {
+			assembly = SelectAssembly (arch, name, entry, loading: true);
+			if (assembly != null) {
+				return assembly;
+			}
+		}
+
+		try {
+			assembly = ReadAssembly (filePath);
+		} catch (Exception e) when (e is FileNotFoundException || e is DirectoryNotFoundException) {
+			// These are ok, we can return null
+			return null;
+		}
+
+		if (!cache.TryGetValue (name, out entry)) {
+			entry = new CacheEntry (log, filePath, assembly, arch);
+			cache.Add (name, entry);
```

Comment:
When `forceLoad:true`, we'll possibly be calling `cache.Add(name, entry)` multiple times, possibly for the same ABI, which will elicit a warning message.  This feels undesirable, but relatedly, I can't find any codepaths which call `Load(…, forceLoad:true)`.

Why support `bool forceLoad` at all if nothing is using it?

Reply:
`cache.Add (name, entry)` is protected by `TryGetValue` and shouldn't throw. Kept `forceLoad` for compatibility, "just in case".